### PR TITLE
Add alert for unscheduled silence sync job

### PR DESCRIPTION
**Pull request by QuentinBisson**
Towards https://github.com/giantswarm/giantswarm/issues/22219

This PR:

- adds an alert for unscheduled silence sync job

### Checklist

- [x] Update changelog in CHANGELOG.md.
- [x] Request review from oncall area, as well as team (e.g: `oncall-kaas-cloud` GitHub group).
- [x] Alerting rules must have a comment documenting why it needs to exist.

### CODE REVIEW

**Review by TheoBrigitte — CHANGES_REQUESTED** [review]

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml [hunk 1]
```diff
@@ -22,6 +22,18 @@ spec:
         severity: notify
         team: {{ include "providerTeam" . }}
         topic: managementcluster
+    ## This alert triggers when the silence sync job did not schedule for more than 2 hours.
+    - alert: SilenceJobHasNotBeenScheduledForTooLong
+      annotations:
+        description: '{{`CronJob {{ $labels.namespace }}/{{ $labels.cronjob }} has not been scheduled for more than 2 hours.`}}'
+        opsrecipe: job-has-not-been-scheduled-for-too-long/
+      expr: (time() - kube_cronjob_status_last_schedule_time{cronjob="silence-operator-sync"} ) > 7200
```
[the remark this answers is not in the document]
Comment (QuentinBisson) [reply helm/prometheus-rules/templates/alerting-rules/job.rules.yml:30]: Let's start with 1 day then

**Comment by QuentinBisson** [comment]
Do we want to make this more generic (i.e. define a criticality for cronjobs on MCs at least that would page if highest after 2 hours to lowest after 1 day) ?

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml [hunk 2]
```diff
@@ -22,6 +22,18 @@ spec:
         severity: notify
         team: {{ include "providerTeam" . }}
         topic: managementcluster
+    ## This alert triggers when the silence sync job did not schedule for more than 1 day.
+    - alert: SilenceJobHasNotBeenScheduledForTooLong
+      annotations:
+        description: '{{`CronJob {{ $labels.namespace }}/{{ $labels.cronjob }} has not been scheduled for more than 1 day.`}}'
+        opsrecipe: job-has-not-been-scheduled-for-too-long/
+      expr: (time() - kube_cronjob_status_last_schedule_time{cronjob="silence-operator-sync"} ) > 86400
```
Comment (TheoBrigitte) [inline helm/prometheus-rules/templates/alerting-rules/job.rules.yml:30]: Does the `kube_cronjob_status_last_schedule_time` metrics tell us that the job actually succeeded ? I feel like it does not.

There is another metric `kube_job_status_succeeded` but it does not tell us when the job ran.

I would suggest we use a combination of both, so we know the job ran in the last day and there was at least 1 success.

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml [hunk 3]
```diff
@@ -22,6 +22,18 @@ spec:
...
+      for: 15m
```
Comment (TheoBrigitte) [inline helm/prometheus-rules/templates/alerting-rules/job.rules.yml:31]: Why do we need this when we are already taking data over 1 day?

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml (hunk 2 above)
Comment (TheoBrigitte) [reply helm/prometheus-rules/templates/alerting-rules/job.rules.yml:30]: Maybe some ideas there https://lyz-code.github.io/blue-book/devops/kubernetes/kubernetes_jobs/#building-our-alert

Comment (QuentinBisson) [reply helm/prometheus-rules/templates/alerting-rules/job.rules.yml]: That's some interesting link

**Comment by QuentinBisson** [comment]
@TheoBrigitte can you take another look now ?

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml [hunk 4]
```diff
@@ -14,14 +14,27 @@ spec:
...
-      expr: kube_job_failed{cluster_type="management_cluster", condition="true"} == 1
+      expr: kube_job_failed{cluster_type="management_cluster", condition="true", job_name!~"silence-operator-sync.+|grafana-permission.+"} == 1
       for: 15m
       labels:
         area: kaas
         cancel_if_outside_working_hours: {{ include "workingHoursOnly" . }}
         severity: notify
         team: {{ include "providerTeam" . }}
         topic: managementcluster
+    - alert: AtlasJobHasNotBeenScheduledForTooLong
```
Comment (TheoBrigitte) [inline helm/prometheus-rules/templates/alerting-rules/job.rules.yml:25]: Can we please not put team name into alert name. We are already using label for this.

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml [hunk 5]
```diff
@@ -14,14 +14,27 @@ spec:
...
+      annotations:
+        description: '{{`CronJob {{ $labels.namespace }}/{{ $labels.cronjob }} has not been scheduled for more than 1 day.`}}'
+        opsrecipe: job-has-not-been-scheduled-for-too-long/
+      # This alert triggers when the silence sync or grafana permission job did not schedule for more than 1 day
+      # or if the job did not run successfully at least once in the last day
+      expr: (time() - kube_cronjob_status_last_schedule_time{cronjob=~"silence-operator-sync|grafana-permissions"} ) > 86400
+            or count(max_over_time(kube_job_status_succeeded{job_name=~"silence-operator-sync.+|grafana-permission.+"}[1d]) == 1) == 0
```
Comment (TheoBrigitte) [inline helm/prometheus-rules/templates/alerting-rules/job.rules.yml:32]: I would rather have 2 different alert for `silence-operator-sync` and `grafana-permission` job.

**Review by TheoBrigitte — APPROVED** [review]
Looks great.

I am really unsure if we should bundle both `silence-operator-sync` and `grafana-permission` job together in this alert. But we can try and change later if needed.

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml (hunk 4 above)
Comment (QuentinBisson) [reply helm/prometheus-rules/templates/alerting-rules/job.rules.yml:25]: Alrighty

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml (hunk 5 above)
Comment (QuentinBisson) [reply helm/prometheus-rules/templates/alerting-rules/job.rules.yml]: alerts are splitted

File: helm/prometheus-rules/templates/alerting-rules/job.rules.yml (hunk 2 above)
Comment (hervenicol) [reply helm/prometheus-rules/templates/alerting-rules/job.rules.yml:31]: Hey, look at the new metric that's coming! :wink: 
https://github.com/kubernetes/kube-state-metrics/pull/1732

Comment (QuentinBisson) [reply helm/prometheus-rules/templates/alerting-rules/job.rules.yml:31]: Interesting :)

**Pull request merged**